### ISSUE TITLE
Random access of mp4 file does not yield consistent frame array for a given index

### ISSUE THREAD

**Issue by kvoit** [imageio/imageio#938]
We are running imageio 2.19.3 on python 3.8. 
We are using reader.get_data(fr_idx) (v2 style) for random access to a frame in an mp4 video. 

Unfortunately, depending on the frame index of the previously requested frame, get_data yields different arrays for the same frame index. 
In v3 style access with imread, it does not seem to depend on the previous frame (as it is stateless?) but the output differs from the output of the iterator.

Here is some v2 test code:
```
import imageio
import hashlib
from tqdm import tqdm

reader = imageio.get_reader("360_0011.MP4")
reader2 = imageio.get_reader("360_0011.MP4")

# Build up a hash library
hashes = dict()
for i_fr, img in enumerate(tqdm(reader)):
    hashes[i_fr] = hashlib.md5(img).hexdigest()

# Query frame 123 after frame 0
fr_idx = 123
reader2.get_data(0)
fr_hash = hashlib.md5(reader2.get_data(fr_idx)).hexdigest()
print(fr_hash == hashes[fr_idx])
print(fr_hash in hashes.values())
list(hashes.values()).index(fr_hash)
>> True
>> True
>> 123

# Query frame 124 after frame 0
fr_idx = 124
reader2.get_data(0)
fr_hash = hashlib.md5(reader2.get_data(fr_idx)).hexdigest()
print(fr_hash == hashes[fr_idx])
print(fr_hash in hashes.values())
list(hashes.values()).index(fr_hash)
>> False
>> True
>> 125

# Query frame 125 after frame 0
fr_idx = 125
reader2.get_data(0)
fr_hash = hashlib.md5(reader2.get_data(fr_idx)).hexdigest()
print(fr_hash == hashes[fr_idx])
print(fr_hash in hashes.values())
list(hashes.values()).index(fr_hash)
>> False
>> True
>> 126

# Query frame 124
fr_idx = 124
fr_hash = hashlib.md5(reader2.get_data(fr_idx)).hexdigest()
print(fr_hash == hashes[fr_idx])
print(fr_hash in hashes.values())
list(hashes.values()).index(fr_hash)
>> False
>> True
>> 125

# Query frame 123
fr_idx = 123
fr_hash = hashlib.md5(reader2.get_data(fr_idx)).hexdigest()
print(fr_hash == hashes[fr_idx])
print(fr_hash in hashes.values())
list(hashes.values()).index(fr_hash)
>> True
>> True
>> 123

# Query frame 124
fr_idx = 124
fr_hash = hashlib.md5(reader2.get_data(fr_idx)).hexdigest()
print(fr_hash == hashes[fr_idx])
print(fr_hash in hashes.values())
list(hashes.values()).index(fr_hash)
>> True
>> True
>> 124
```

I would prefer to avoid posting the video publicly, but I would be happy to provide it to anyone in direct communication who would like to take the issue on.

**Comment by FirefoxMetzger** [comment]
I ran your script on one of our test videos and it all worked as it should. It could be a video-specific or instance-specific problem. Could you run the snippet below (it will run your script on our test video) on your machine and see if it behaves incorrectly?

<details>

```python
import imageio
import hashlib

video = "imageio:cockatoo.mp4"
reader = imageio.get_reader(video)
reader2 = imageio.get_reader(video)

# Build up a hash library
hashes = dict()
for i_fr, img in enumerate(reader):
    hashes[i_fr] = hashlib.md5(img).hexdigest()

# Query frame 123 after frame 0
fr_idx = 123
reader2.get_data(0)
fr_hash = hashlib.md5(reader2.get_data(fr_idx)).hexdigest()
print(fr_hash == hashes[fr_idx])
print(fr_hash in hashes.values())
print(list(hashes.values()).index(fr_hash))
print("--")

# Query frame 124 after frame 0
fr_idx = 124
reader2.get_data(0)
fr_hash = hashlib.md5(reader2.get_data(fr_idx)).hexdigest()
print(fr_hash == hashes[fr_idx])
print(fr_hash in hashes.values())
print(list(hashes.values()).index(fr_hash))
print("--")

# Query frame 125 after frame 0
fr_idx = 125
reader2.get_data(0)
fr_hash = hashlib.md5(reader2.get_data(fr_idx)).hexdigest()
print(fr_hash == hashes[fr_idx])
print(fr_hash in hashes.values())
print(list(hashes.values()).index(fr_hash))
print("--")

# Query frame 124
fr_idx = 124
fr_hash = hashlib.md5(reader2.get_data(fr_idx)).hexdigest()
print(fr_hash == hashes[fr_idx])
print(fr_hash in hashes.values())
print(list(hashes.values()).index(fr_hash))
print("--")

# Query frame 123
fr_idx = 123
fr_hash = hashlib.md5(reader2.get_data(fr_idx)).hexdigest()
print(fr_hash == hashes[fr_idx])
print(fr_hash in hashes.values())
print(list(hashes.values()).index(fr_hash))
print("--")

# Query frame 124
fr_idx = 124
fr_hash = hashlib.md5(reader2.get_data(fr_idx)).hexdigest()
print(fr_hash == hashes[fr_idx])
print(fr_hash in hashes.values())
print(list(hashes.values()).index(fr_hash))
print("--")

```

</details>

In the meantime, you can send me the video if you'd like and I can check if it is video-specific and (if so) check if I can figure out why it behaves this way. (sebastian@wallkoetter.net)

**Comment by kvoit** [comment]
Thanks for looking into this. I have provided you 2 videos via mail. For one, it seems to be sufficient to read the previous frame, for the other, it isn't.

The test video is extracted just fine, all returns are True.

**Comment by kvoit** [comment]
I have made another comparison to [PIMS PyAVReaderTimed](https://soft-matter.github.io/pims/dev/video.html). PIMS seems to manage to yield consistent frames with both iterator access and random access (get_frame()).
What I did notice is that the hash list that PIMS generates is different. At some point, it returns a frame twice, and does not return the last frame at all:

From cam1.mp4:
```
for n in range(620,625):
    print(f"{n} {hashes_pims[n]} {hashes_imageio[n]}")
620 fce4da92e07f132e11bdee7b44322b6a fce4da92e07f132e11bdee7b44322b6a
621 2e43a6209ad4d87a3435f5d0d9a5d5d5 2e43a6209ad4d87a3435f5d0d9a5d5d5
622 2e43a6209ad4d87a3435f5d0d9a5d5d5 c7866a3a91355e487192901951714c87
623 c7866a3a91355e487192901951714c87 7a219477137e364aa832f3f03fbea151
624 7a219477137e364aa832f3f03fbea151 e29e4fb648d37c5c99a876ca452e654b
```

```
for n in range(990,995):
    print(f"{n} {hashes_pims[n]} {hashes_imageio[n]}")
990 35a18dbe59f28598b36a2a9a7ca082ae 6e72e375f58b1fe6d7c9397ca8fe0ac2
991 6e72e375f58b1fe6d7c9397ca8fe0ac2 3771c6599ad9147be24f897c69b3fb3b
992 3771c6599ad9147be24f897c69b3fb3b 7ffc6d7c2ff353fee0a85ea46925bc7f
993 7ffc6d7c2ff353fee0a85ea46925bc7f 2a8648af9e3eb0a342594dedeb49d8af
994 2a8648af9e3eb0a342594dedeb49d8af 27a18c5d62e707fbb65223e62157abc4
```

I wouldn't necessarily say that is better, but perhaps it helps getting to the root of issues ... I guess a full solution should give consistent frames and access to all of them :)

**Comment by FirefoxMetzger** [comment]
I just tried the following with your video (using the `pyav` backend instead of the `imageio-ffmpeg` one), and it appears to read frames correctly.

```python
import hashlib
from tqdm import tqdm
import imageio.v3 as iio

filename = "360_0011.MP4"

# Build up a hash library
n_frames = iio.improps(filename, plugin="pyav").shape[0]
hashes = dict()
for i_fr, img in enumerate(tqdm(iio.imiter(filename, plugin="pyav", thread_type="FRAME", thread_count=16), total=n_frames)):
    hashes[i_fr] = hashlib.md5(img).hexdigest()

with iio.imopen(filename, "r", plugin="pyav") as file:
    for idx in [123, 124, 125]:
        file.read(index=0)
        img = file.read(index=idx)
        fr_hash = hashlib.md5(img).hexdigest()
        print(fr_hash == hashes[idx])
        print(fr_hash in hashes.values())
        print(list(hashes.values()).index(fr_hash))

    for idx in [124, 123, 124]:
        fr_hash = hashlib.md5(file.read(index=idx)).hexdigest()
        print(fr_hash == hashes[idx])
        print(fr_hash in hashes.values())
        print(list(hashes.values()).index(fr_hash))

```

```
100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 4866/4866 [01:16<00:00, 63.64it/s]
True
True
123
True
True
124
True
True
125
True
True
124
True
True
123
True
True
124
```

I also tried the same code using the `imageio-ffmpeg` backend. In this case, I can indeed reproduce the problem:

```python
import hashlib
from tqdm import tqdm
import imageio.v3 as iio

filename = "360_0011.MP4"

# Build up a hash library
n_frames = iio.improps(filename, plugin="pyav").shape[0]
hashes = dict()
for i_fr, img in enumerate(tqdm(iio.imiter(filename, plugin="pyav", thread_type="FRAME", thread_count=16), total=n_frames)):
    hashes[i_fr] = hashlib.md5(img).hexdigest()

with iio.imopen(filename, "r", plugin="FFMPEG") as file:
    for idx in [123, 124, 125]:
        file.read(index=0)
        img = file.read(index=idx)
        fr_hash = hashlib.md5(img).hexdigest()
        print(fr_hash == hashes[idx])
        print(fr_hash in hashes.values())
        print(list(hashes.values()).index(fr_hash))

    for idx in [124, 123, 124]:
        fr_hash = hashlib.md5(file.read(index=idx)).hexdigest()
        print(fr_hash == hashes[idx])
        print(fr_hash in hashes.values())
        print(list(hashes.values()).index(fr_hash))
```

```
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 4866/4866 [01:17<00:00, 62.45it/s]
True
True
123
False
True
125
False
True
126
False
True
125
True
True
123
False
True
125
```

**Comment by kvoit** [comment]
Thanks for that. That certainly helps a lot. 

Yes, in principle we can change over to v3.
However, we have a custom filter for a legacy proprietary video format:
https://github.com/bbo-lab/ccvtools/tree/main/ccvtools/rawio
With naive usage of v2, it was automatically used once rawio was imported from ccvtools.
When I explicitly specify pyav, then it does not seem to work. Is there an intended way to have this again, without having to check the video type outside imageio manually?

**Comment by kvoit** [comment]
Thanks a lot, I think we can work with this. Further tests with v3 and pyav showed no issues by now. I'll keep a wrapper class that checks hashes vor a while, though.

**Comment by kvoit** [comment]
Ah, beans ...

It doesn't read the last frame with random access ... it DOES read it with the iterator.

This is your code, only that I changes to cam1.mp4, and added frame_idx 994 (of 995 frames) to the list:
```
import hashlib
from tqdm import tqdm
import imageio.v3 as iio

filename = "cam1.mp4"

# Build up a hash library
n_frames = iio.improps(filename, plugin="pyav").shape[0]
hashes = dict()
for i_fr, img in enumerate(tqdm(iio.imiter(filename, plugin="pyav", thread_type="FRAME", thread_count=16), total=n_frames)):
    hashes[i_fr] = hashlib.md5(img).hexdigest()

with iio.imopen(filename, "r", plugin="pyav") as file:
    for idx in [123, 124, 125, 994]:
        file.read(index=0)
        img = file.read(index=idx)
        fr_hash = hashlib.md5(img).hexdigest()
        print(fr_hash == hashes[idx])
        print(fr_hash in hashes.values())
        print(list(hashes.values()).index(fr_hash))

    for idx in [124, 123, 124]:
        fr_hash = hashlib.md5(file.read(index=idx)).hexdigest()
        print(fr_hash == hashes[idx])
        print(fr_hash in hashes.values())
        print(list(hashes.values()).index(fr_hash))
```

**Comment by FirefoxMetzger** [comment]
@kvoit Which branch are you on? For me, it iterates correctly on `master`; however, in #939 it misses the last frame (one of the issues that still needs fixing before it gets merged).

**Comment by kvoit** [comment]
That was it, thanks. Yes, I still had the pull request repo there. Sorry!

**Comment by kvoit** [comment]
I just noticed accessing any frame higher up takes almost 2 minutes, even consecutive ones? :(

```
import imageio.v3 as iio
import hashlib
import time

file_name = "/media/smb/soma-fs.ad01.caesar.de/bbo/analysis/pose/data/mp4/20220106_imutest-teensy_0011/360_0011.MP4"
plugin = "pyav"
with iio.imopen(file_name, "r", plugin=plugin) as reader:
    tic = time.perf_counter()
    img = reader.read(index=3600)
    toc = time.perf_counter()
    print(f"read frame in {toc - tic:0.4f} seconds")

    tic = time.perf_counter()
    img = reader.read(index=3602)
    toc = time.perf_counter()
    print(f"read frame in {toc - tic:0.4f} seconds")

    tic = time.perf_counter()
    img = reader.read(index=0)
    toc = time.perf_counter()
    print(f"read frame in {toc - tic:0.4f} seconds")

    tic = time.perf_counter()
    img = reader.read(index=3600)
    toc = time.perf_counter()
    print(f"read frame in {toc - tic:0.4f} seconds")

    tic = time.perf_counter()
    img = reader.read(index=3601)
    toc = time.perf_counter()
    print(f"read frame in {toc - tic:0.4f} seconds")
```

```
read frame in 103.1166 seconds
read frame in 102.2579 seconds
read frame in 0.0351 seconds
read frame in 104.2764 seconds
read frame in 103.1596 seconds
```

**Comment by FirefoxMetzger** [comment]
> I just noticed accessing any frame higher up takes almost 2 minutes, even consecutive ones? :(

Yes, this is one of the things that the PR (#939) addresses. Without getting hung up on the details of why (TL:DR pyAV's docs are not good), you will find that read times for consecutive reads will improve significantly in #939 

```
read frame in 0.0550 seconds
read frame in 0.0909 seconds
read frame in 0.0505 seconds
read frame in 0.0533 seconds
read frame in 0.0719 seconds
```

With a bit of luck I will find the time to finalize the PR this week, in which case it will roll out next Monday.

**Comment by FirefoxMetzger** [comment]
#939 Is now merged. It fixes this behavior, so I will close the issue as resolved. The fix will roll out next Monday (v2.26.0).

**Comment by kvoit** [comment]
This indeed seems to have solved all issues. Thanks a lot, again!
